# Introduction To Data Science HW9 Chi-squared, T-test, Linear Regression

## Chi-squared

For example, in the below is student smoke and exercise relationship. As for Exercise, they are "Freq" (frequently), "Some" and "None". As for Smoke, they are "Heavy", "Regul" (regularly), "Occas" (occasionally) and "Never".

|        | Freq  | None  | Some  |
|--------|-------|-------|-------|
| Heavy  |   7   |   1   |   3   |
| Never  |  87   |  18   |  84   |
| Occas  |  12   |   3   |   4   |
| Regul  |   9   |   1   |   7   |

### Question: 
Test the hypothesis whether the students smoking habit is independent of their exercise level at .05 significance level

In [30]:
import numpy as np
import pandas as pd
import scipy.stats as stats

In [32]:
# First create the survey list for the table
survey = [ [ 7, 1, 3 ], [ 87, 18, 84 ], [ 12, 3, 4 ], [ 9, 1, 7 ] ]
# Later use 
chi2, p, ddof, expected = stats.chi2_contingency( survey )
msg = "Test Statistic: {}\np-value: {}\nDegrees of Freedom: {}\n"
print( msg.format( chi2, p, ddof ) )
print( expected )

Test Statistic: 5.488545890584232
p-value: 0.48284216946545633
Degrees of Freedom: 6

[[  5.36016949   1.0720339    4.56779661]
 [ 92.09745763  18.41949153  78.48305085]
 [  9.25847458   1.85169492   7.88983051]
 [  8.28389831   1.65677966   7.05932203]]


###  Interpretation:
As the p-value 0.4828 is greater than the .05 significance level, we do not reject the null hypothesis that the smoking habit is independent of the exercise level of the students. Also for the test statistics 5.48 is not very large.

## T-test

In [38]:
# First you have pip install rpy2, then you can use rpy2 to get datasets from r
from rpy2.robjects import r, pandas2ri
pandas2ri.activate()

In [45]:
# Get the dataset iris using rpy2 from r 
iris= r['iris']
# Show the first 5 rows of iris
iris.head()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species
1           5.1          3.5           1.4          0.2  setosa
2           4.9          3.0           1.4          0.2  setosa
3           4.7          3.2           1.3          0.2  setosa
4           4.6          3.1           1.5          0.2  setosa
5           5.0          3.6           1.4          0.2  setosa

### Question: 
We want to know Sepal.Length has some affect on Species or not

In [64]:
# Create two array for setosa and versicolor their Sepal.Length values
setosa= iris[iris.Species== 'setosa']
s= setosa['Sepal.Length']
versicolor= iris[iris.Species== 'versicolor']
v= versicolor['Sepal.Length']
# Perform two sample t-test
stats.ttest_ind(a= s, b= v,equal_var=False)

Ttest_indResult(statistic=-10.520986267549111, pvalue=3.7467426139838419e-17)

###  Interpretation:
Since pvalue is 3.7467426139838419e-17 much smaller than 0.05, we can reject the h0 hypotheis, that Sepal.Length can be a good predictor for iris classfication.

## Linear Regression

In [81]:
# Use sklearn dataset boston house market to pridct house price
from sklearn import datasets
boston = datasets.load_boston()
from sklearn.linear_model import LinearRegression

In [74]:
# Show the detail about boston dataset
print (boston.DESCR)

Boston House Prices dataset

Notes
------
Data Set Characteristics:  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive
    
    :Median Value (attribute 14) is usually the target

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pupil-teacher ratio by town
      

In [77]:
# First convert sklearn boston dataset into panda dataframe
bos= pd.DataFrame(boston.data, columns=boston.feature_names)
# Show first 5 rows of bos dataset
bos.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [80]:
# Add one column Price( the variable we want to predict) in bos dataset as target
bos['Price']= boston.target
bos.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  Price  
0     15.3  396.90   4.98   24.0  
1     17.8  396.90   9.14   21.6  
2     17.8  392.83   4.03   34.7  
3     18.7  394.63   2.94   33.4  
4     18.7  396.90   5.33   36.2

### Question: 
We want to know each paremeter their relationship to the house price

In [89]:
# Set up X,Y for linear regression, X is all variables without price, Y is price
X= bos.drop('Price', axis= 1)
Y= bos['Price']
# Create LinearRegression object
lm= LinearRegression()
# Fit the linear model
lm.fit(X,Y)
# Show the R square 
lm.score(X,Y)

0.7406077428649428

In [109]:
# Create coefficinet dataframe of each parameter
summary=pd.DataFrame(list(zip(X.columns, lm.coef_)), columns=['feature', 'coefficient'])
summary

feature  coefficient
0      CRIM    -0.107171
1        ZN     0.046395
2     INDUS     0.020860
3      CHAS     2.688561
4       NOX   -17.795759
5        RM     3.804752
6       AGE     0.000751
7       DIS    -1.475759
8       RAD     0.305655
9       TAX    -0.012329
10  PTRATIO    -0.953464
11        B     0.009393
12    LSTAT    -0.525467

In [111]:
# Also show the intercept for this linear regression
lm.intercept_

36.491103280363404

###  Interpretation:
First check the R squared 0.74 is not too small, so we can use this model, then fit the linear regression we can get the Y=coefficient_{i}*feature_{i}+intercept 

## Reference
1. http://connor-johnson.com/2014/12/31/the-pearson-chi-squared-test-with-python-and-r/
2. https://docs.scipy.org/doc/scipy-0.19.1/reference/generated/scipy.stats.ttest_ind.html
3. https://pandas.pydata.org/pandas-docs/stable/r_interface.html
4. http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html